File: phases/08-compilers-and-language-design/02-lexing-i-regex-dfa-lexer/code/main.py
"""Lesson 02: Regex-based Lexer — Python implementation.

A token-stream lexer that converts source text into a sequence of
classified tokens. Supports identifiers, keywords, integers, floats,
strings with escape sequences, operators, punctuation, and comments.
"""

from dataclasses import dataclass
from enum import Enum, auto
from typing import Optional


class TokenKind(Enum):
    KEYWORD = auto()
    IDENT = auto()
    INT = auto()
    FLOAT = auto()
    STRING = auto()
    OP = auto()
    LPAREN = auto()
    RPAREN = auto()
    LBRACE = auto()
    RBRACE = auto()
    SEMI = auto()
    COMMA = auto()
    ASSIGN = auto()
    EOF = auto()
    ERROR = auto()


@dataclass
class Token:
    kind: TokenKind
    value: str = ""
    line: int = 0
    col: int = 0

    def __str__(self) -> str:
        if self.kind == TokenKind.ERROR:
            return f"ERROR({self.line}:{self.col}: {self.value})"
        if self.kind == TokenKind.EOF:
            return "EOF"
        if self.value:
            return f"{self.kind.name}({self.value})"
        return self.kind.name


KEYWORDS = frozenset([
    "if", "else", "while", "for", "return", "fn", "let", "mut", "true", "false",
])


class Lexer:
    def __init__(self, source: str) -> None:
        self._chars = list(source)
        self._pos = 0
        self._line = 1
        self._col = 1

    # ------------------------------------------------------------------ helpers

    def _current(self) -> Optional[str]:
        if self._pos < len(self._chars):
            return self._chars[self._pos]
        return None

    def _peek(self) -> Optional[str]:
        nxt = self._pos + 1
        if nxt < len(self._chars):
            return self._chars[nxt]
        return None

    def _advance(self) -> Optional[str]:
        ch = self._current()
        if ch is None:
            return None
        self._pos += 1
        if ch == "\n":
            self._line += 1
            self._col = 1
        else:
            self._col += 1
        return ch

    # ------------------------------------------------------------------ skip

    def _skip_whitespace(self) -> None:
        while True:
            ch = self._current()
            if ch is None:
                return
            if ch.isspace():
                self._advance()
            elif ch == "/" and self._peek() == "/":
                while True:
                    c = self._advance()
                    if c is None or c == "\n":
                        break
            else:
                return

    # ------------------------------------------------------------------ scanners

    def _scan_identifier(self) -> Token:
        line, col = self._line, self._col
        name_chars: list[str] = []
        while True:
            ch = self._current()
            if ch is not None and (ch.isalnum() or ch == "_"):
                name_chars.append(ch)
                self._advance()
            else:
                break
        name = "".join(name_chars)
        kind = TokenKind.KEYWORD if name in KEYWORDS else TokenKind.IDENT
        return Token(kind, name, line, col)

    def _scan_number(self) -> Token:
        line, col = self._line, self._col
        parts: list[str] = []

        # prefix detection for hex / bin / oct
        if self._current() == "0":
            parts.append("0")
            self._advance()
            nxt = self._current()
            if nxt in ("x", "X"):
                parts.append(self._advance() or "")  # type: ignore[arg-type]
                while True:
                    ch = self._current()
                    if ch is not None and ch.isascii() and ch.isalnum():
                        parts.append(self._advance() or "")  # type: ignore[arg-type]
                    else:
                        break
                try:
                    int("".join(parts)[2:], 16)
                except ValueError:
                    return Token(TokenKind.ERROR, f"invalid hex literal: {''.join(parts)}", line, col)
                return Token(TokenKind.INT, "".join(parts), line, col)
            if nxt in ("b", "B"):
                parts.append(self._advance() or "")  # type: ignore[arg-type]
                while True:
                    ch = self._current()
                    if ch in ("0", "1"):
                        parts.append(self._advance() or "")  # type: ignore[arg-type]
                    else:
                        break
                return Token(TokenKind.INT, "".join(parts), line, col)
            if nxt in ("o", "O"):
                parts.append(self._advance() or "")  # type: ignore[arg-type]
                while True:
                    ch = self._current()
                    if ch is not None and "0" <= ch <= "7":
                        parts.append(self._advance() or "")  # type: ignore[arg-type]
                    else:
                        break
                return Token(TokenKind.INT, "".join(parts), line, col)

        # decimal digits
        while True:
            ch = self._current()
            if ch is not None and ch.isdigit():
                parts.append(self._advance() or "")  # type: ignore[arg-type]
            else:
                break

        # float
        if self._current() == "." and (self._peek() or "").isdigit():
            parts.append(self._advance() or "")  # type: ignore[arg-type]
            while True:
                ch = self._current()
                if ch is not None and ch.isdigit():
                    parts.append(self._advance() or "")  # type: ignore[arg-type]
                else:
                    break
            return Token(TokenKind.FLOAT, "".join(parts), line, col)

        return Token(TokenKind.INT, "".join(parts), line, col)

    def _scan_string(self, quote: str) -> Token:
        line, col = self._line, self._col
        self._advance()  # opening quote
        buf: list[str] = []
        while True:
            ch = self._current()
            if ch is None:
                return Token(TokenKind.ERROR, "unterminated string literal", line, col)
            if ch == quote:
                self._advance()  # closing quote
                return Token(TokenKind.STRING, "".join(buf), line, col)
            if ch == "\\":
                self._advance()
                esc = self._current()
                escapes = {"n": "\n", "t": "\t", "r": "\r", "\\": "\\", '"': '"', "'": "'", "0": "\0"}
                if esc in escapes:
                    buf.append(escapes[esc])
                    self._advance()
                elif esc is not None:
                    buf.append("\\")
                    buf.append(esc)
                    self._advance()
                else:
                    return Token(TokenKind.ERROR, "unterminated string literal", line, col)
            else:
                buf.append(ch)
                self._advance()

    def _scan_operator(self) -> Token:
        ch = self._advance()
        op = ch or ""
        nxt = self._current()
        two_char = {"==", "!=", "<=", ">=", "&&", "||"}
        if nxt is not None and (op + nxt) in two_char:
            op += self._advance()  # type: ignore[operator]
        return Token(TokenKind.OP, op, self._line, self._col)

    # ------------------------------------------------------------------ public

    def next_token(self) -> Token:
        self._skip_whitespace()
        ch = self._current()
        if ch is None:
            return Token(TokenKind.EOF, "", self._line, self._col)
        if ch.isalpha() or ch == "_":
            return self._scan_identifier()
        if ch.isdigit():
            return self._scan_number()
        if ch == '"':
            return self._scan_string('"')
        if ch == "'":
            return self._scan_string("'")
        simple = {
            "(": TokenKind.LPAREN,
            ")": TokenKind.RPAREN,
            "{": TokenKind.LBRACE,
            "}": TokenKind.RBRACE,
            ";": TokenKind.SEMI,
            ",": TokenKind.COMMA,
        }
        if ch in simple:
            self._advance()
            return Token(simple[ch], "", self._line, self._col)
        if ch == "=" and self._peek() != "=":
            self._advance()
            return Token(TokenKind.ASSIGN, "", self._line, self._col)
        if ch in "+-*/=!<>|&":
            return self._scan_operator()
        self._advance()
        return Token(TokenKind.ERROR, f"unexpected character: '{ch}'", self._line, self._col)

    def tokenize(self) -> list[Token]:
        tokens: list[Token] = []
        while True:
            tok = self.next_token()
            tokens.append(tok)
            if tok.kind == TokenKind.EOF:
                break
        return tokens


# --------------------------------------------------------------------------- main

def main() -> None:
    source = '''
fn main() {
    let x = 42 + 0xFF;
    let pi = 3.14;
    let name = "hello\\nworld";
    // This is a comment
    if x == 100 {
        return true;
    }
}
'''
    lexer = Lexer(source)
    for tok in lexer.tokenize():
        print(tok)


if __name__ == "__main__":
    main()

File: phases/08-compilers-and-language-design/02-lexing-i-regex-dfa-lexer/code/main.rs
use std::fmt;

#[derive(Debug, Clone, PartialEq)]
pub enum Token {
    // Keywords
    Keyword(String),
    // Identifiers
    Ident(String),
    // Numbers
    Int(i64),
    Float(f64),
    // Strings
    String(String),
    // Operators
    Op(String),
    // Punctuation
    LParen,
    RParen,
    LBrace,
    RBrace,
    Semi,
    Comma,
    Assign,
    // Special
    Eof,
    // Error
    Error { msg: String, line: usize, col: usize },
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Token::Keyword(kw) => write!(f, "KEYWORD({})", kw),
            Token::Ident(name) => write!(f, "IDENT({})", name),
            Token::Int(n) => write!(f, "INT({})", n),
            Token::Float(n) => write!(f, "FLOAT({})", n),
            Token::String(s) => write!(f, "STRING({:?})", s),
            Token::Op(op) => write!(f, "OP({})", op),
            Token::LParen => write!(f, "LPAREN"),
            Token::RParen => write!(f, "RPAREN"),
            Token::LBrace => write!(f, "LBRACE"),
            Token::RBrace => write!(f, "RBRACE"),
            Token::Semi => write!(f, "SEMI"),
            Token::Comma => write!(f, "COMMA"),
            Token::Assign => write!(f, "ASSIGN"),
            Token::Eof => write!(f, "EOF"),
            Token::Error { msg, line, col } => write!(f, "ERROR({}:{}: {})", line, col, msg),
        }
    }
}

const KEYWORDS: &[&str] = &[
    "if", "else", "while", "for", "return", "fn", "let", "mut", "true", "false",
];

pub struct Lexer {
    input: Vec<char>,
    pos: usize,
    line: usize,
    col: usize,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        Lexer {
            input: input.chars().collect(),
            pos: 0,
            line: 1,
            col: 1,
        }
    }

    fn current(&self) -> Option<char> {
        self.input.get(self.pos).copied()
    }

    fn peek(&self) -> Option<char> {
        self.input.get(self.pos + 1).copied()
    }

    fn advance(&mut self) -> Option<char> {
        let ch = self.current()?;
        self.pos += 1;
        if ch == '\n' {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        Some(ch)
    }

    fn skip_whitespace(&mut self) {
        while let Some(ch) = self.current() {
            if ch.is_whitespace() {
                self.advance();
            } else if ch == '/' && self.peek() == Some('/') {
                // Single-line comment
                while let Some(c) = self.advance() {
                    if c == '\n' {
                        break;
                    }
                }
            } else {
                break;
            }
        }
    }

    fn scan_identifier(&mut self) -> Token {
        let start_line = self.line;
        let start_col = self.col;
        let mut name = String::new();
        while let Some(ch) = self.current() {
            if ch.is_alphanumeric() || ch == '_' {
                name.push(ch);
                self.advance();
            } else {
                break;
            }
        }
        if KEYWORDS.contains(&name.as_str()) {
            Token::Keyword(name)
        } else {
            Token::Ident(name)
        }
    }

    fn scan_number(&mut self) -> Token {
        let mut num_str = String::new();
        let start_line = self.line;

        // Check for hex, octal, binary prefixes
        if self.current() == Some('0') {
            num_str.push('0');
            self.advance();
            match self.current() {
                Some('x') | Some('X') => {
                    num_str.push('x');
                    self.advance();
                    while let Some(ch) = self.current() {
                        if ch.is_ascii_hexdigit() {
                            num_str.push(ch);
                            self.advance();
                        } else {
                            break;
                        }
                    }
                    return match i64::from_str_radix(&num_str[2..], 16) {
                        Ok(n) => Token::Int(n),
                        Err(_) => Token::Error {
                            msg: format!("invalid hex literal: {}", num_str),
                            line: start_line,
                            col: self.col,
                        },
                    };
                }
                Some('b') | Some('B') => {
                    num_str.push('b');
                    self.advance();
                    while let Some(ch) = self.current() {
                        if ch == '0' || ch == '1' {
                            num_str.push(ch);
                            self.advance();
                        } else {
                            break;
                        }
                    }
                    return match i64::from_str_radix(&num_str[2..], 2) {
                        Ok(n) => Token::Int(n),
                        Err(_) => Token::Error {
                            msg: format!("invalid binary literal: {}", num_str),
                            line: start_line,
                            col: self.col,
                        },
                    };
                }
                Some('o') | Some('O') => {
                    num_str.push('o');
                    self.advance();
                    while let Some(ch) = self.current() {
                        if ('0'..='7').contains(&ch) {
                            num_str.push(ch);
                            self.advance();
                        } else {
                            break;
                        }
                    }
                    return match i64::from_str_radix(&num_str[2..], 8) {
                        Ok(n) => Token::Int(n),
                        Err(_) => Token::Error {
                            msg: format!("invalid octal literal: {}", num_str),
                            line: start_line,
                            col: self.col,
                        },
                    };
                }
                _ => {}
            }
        }

        // Decimal digits
        while let Some(ch) = self.current() {
            if ch.is_ascii_digit() {
                num_str.push(ch);
                self.advance();
            } else {
                break;
            }
        }

        // Float: decimal point followed by digits
        if self.current() == Some('.') && self.peek().map_or(false, |c| c.is_ascii_digit()) {
            num_str.push('.');
            self.advance();
            while let Some(ch) = self.current() {
                if ch.is_ascii_digit() {
                    num_str.push(ch);
                    self.advance();
                } else {
                    break;
                }
            }
            return match num_str.parse::<f64>() {
                Ok(n) => Token::Float(n),
                Err(_) => Token::Error {
                    msg: format!("invalid float literal: {}", num_str),
                    line: start_line,
                    col: self.col,
                },
            };
        }

        match num_str.parse::<i64>() {
            Ok(n) => Token::Int(n),
            Err(_) => Token::Error {
                msg: format!("invalid integer literal: {}", num_str),
                line: start_line,
                col: self.col,
            },
        }
    }

    fn scan_string(&mut self, quote: char) -> Token {
        let start_line = self.line;
        let start_col = self.col;
        self.advance(); // consume opening quote
        let mut s = String::new();
        loop {
            match self.current() {
                None => {
                    return Token::Error {
                        msg: "unterminated string literal".to_string(),
                        line: start_line,
                        col: start_col,
                    };
                }
                Some(ch) if ch == quote => {
                    self.advance(); // consume closing quote
                    return Token::String(s);
                }
                Some('\\') => {
                    self.advance();
                    match self.current() {
                        Some('n') => { s.push('\n'); self.advance(); }
                        Some('t') => { s.push('\t'); self.advance(); }
                        Some('r') => { s.push('\r'); self.advance(); }
                        Some('\\') => { s.push('\\'); self.advance(); }
                        Some('"') => { s.push('"'); self.advance(); }
                        Some('\'') => { s.push('\''); self.advance(); }
                        Some('0') => { s.push('\0'); self.advance(); }
                        Some(c) => {
                            s.push('\\');
                            s.push(c);
                            self.advance();
                        }
                        None => {
                            return Token::Error {
                                msg: "unterminated string literal".to_string(),
                                line: start_line,
                                col: start_col,
                            };
                        }
                    }
                }
                Some(ch) => {
                    s.push(ch);
                    self.advance();
                }
            }
        }
    }

    fn scan_operator(&mut self) -> Token {
        let mut op = String::new();
        let ch = self.advance().unwrap();
        op.push(ch);
        // Two-character operators
        match (ch, self.current()) {
            ('=', Some('=')) | ('!', Some('=')) | ('<', Some('=')) | ('>', Some('=')) => {
                op.push(self.advance().unwrap());
            }
            ('&', Some('&')) | ('|', Some('|')) => {
                op.push(self.advance().unwrap());
            }
            _ => {}
        }
        Token::Op(op)
    }

    pub fn next_token(&mut self) -> Token {
        self.skip_whitespace();
        match self.current() {
            None => Token::Eof,
            Some(ch) if ch.is_alphabetic() || ch == '_' => self.scan_identifier(),
            Some(ch) if ch.is_ascii_digit() => self.scan_number(),
            Some('"') => self.scan_string('"'),
            Some('\'') => self.scan_string('\''),
            Some('(') => { self.advance(); Token::LParen }
            Some(')') => { self.advance(); Token::RParen }
            Some('{') => { self.advance(); Token::LBrace }
            Some('}') => { self.advance(); Token::RBrace }
            Some(';') => { self.advance(); Token::Semi }
            Some(',') => { self.advance(); Token::Comma }
            Some('=') if self.peek() != Some('=') => { self.advance(); Token::Assign }
            Some('+') | Some('-') | Some('*') | Some('/') |
            Some('=') | Some('!') | Some('<') | Some('>') |
            Some('&') | Some('|') => self.scan_operator(),
            Some(ch) => {
                let line = self.line;
                let col = self.col;
                self.advance();
                Token::Error {
                    msg: format!("unexpected character: '{}'", ch),
                    line,
                    col,
                }
            }
        }
    }

    pub fn tokenize(&mut self) -> Vec<Token> {
        let mut tokens = Vec::new();
        loop {
            let token = self.next_token();
            let is_eof = matches!(token, Token::Eof);
            tokens.push(token);
            if is_eof {
                break;
            }
        }
        tokens
    }
}

fn main() {
    let source = r#"
fn main() {
    let x = 42 + 0xFF;
    let pi = 3.14;
    let name = "hello\nworld";
    // This is a comment
    if x == 100 {
        return true;
    }
}
"#;

    let mut lexer = Lexer::new(source);
    let tokens = lexer.tokenize();

    for token in &tokens {
        println!("{}", token);
    }
}
